[Proposal] SOAR-0011: Improved error handling (#626)
apple/swift-openapi-generator: Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0011.md

# SOAR-0011: Improved Error Handling

Improve error handling by adding the ability for mapping application errors to HTTP responses.

## Overview

- Proposal: SOAR-0011
- Author(s): [Gayathri Sairamkrishnan](https://github.com/gayathrisairam)

> - Status: **Awaiting Review**

Reviewer: ```suggestion
- Status: **In Review**
```

- Issue: [apple/swift-openapi-generator#609](https://github.com/apple/swift-openapi-generator/issues/609)
- Affected components:
    - runtime

### Introduction

The goal of this proposal to improve the current error handling mechanism in Swift OpenAPI runtime. The proposal 
introduces a way for users to map errors thrown by their handlers to specific HTTP responses.

### Motivation

>  When implementing a server with Swift OpenAPI Generator, users implement a type that conforms to a generated protocol, providing one method for each API operation defined in the OpenAPI document. At runtime, if this function throws, the runtime library transforms this into a 500 HTTP response (Internal Error).

Reviewer: ```suggestion
 When implementing a server with Swift OpenAPI Generator, users implement a type that conforms to a generated protocol, providing one method for each API operation defined in the OpenAPI document. At runtime, if this function throws, it's up to the server transport to transform it into an HTTP response status code – for example, some transport use `500 Internal Error`.
```

Author: Even if the server transport transforms it to an error other than 500, OpenAPI eventually wraps it with 500

Reviewer: Hmm, can you link to where in code that happens? I'm not sure I follow

Author: https://github.com/apple/swift-openapi-runtime/blob/main/Sources/OpenAPIRuntime/Interface/UniversalServer.swift#L115

Reviewer: That's where we wrap errors, but I don't think we ever choose the HTTP response code 500 in swift-openapi-runtime, that's purely a decision of the transport I believe.

Author: Oh yes, you are right. It's Vapor that transforms any error that doesn't confirm to AbortError to 500.

Instead, server developers may want to map errors thrown by the application to a more specific HTTP response. 
Currently, this can be achieved by checking for each error type in each handler's catch block, converting it to an 
appropriate HTTP response and returning it. 

For example,
```swift
func getGreeting(_ input: Operations.getGreeting.Input) async throws -> Operations.getGreeting.Output {
  do {
      let response = try callGreetingLib()
      return .ok(.init(body: response))
  } catch let error {
    switch error {
      case GreetingError.authorizationError:

>         return (HTTPResponse(status: 404), nil)

Reviewer: This needs to be a case from the `Operations.getGreeting.Output`, so e.g. `.notFound(.init())` here? Or since this is `GreetingError.authorizationError`, maybe using `.unauthorized(.init())` is more illustrative?

      case GreetingError.timeout:
        return ...
    }
  }  
}
```
If a user wishes to map many errors, the error handling block scales linearly and introduces a lot of ceremony.

### Proposed solution

The proposed solution is twofold.

1. Provide protocol(s) in `OpenAPIRuntime` to allow users to extend their error types with mappings to HTTP responses.

2. Provide an (opt-in) middleware in OpenAPIRuntime that will call the conversion function on conforming error types when
constructing the HTTP response.

Vapor has a similar mechanism called [AbortError](https://docs.vapor.codes/basics/errors/). 

> HummingBird also has an [error handling mechanism](https://docs.hummingbird.codes/2.0/documentation/hummingbird/errorhandling/)

Reviewer: ```suggestion
Hummingbird also has an [error handling mechanism](https://docs.hummingbird.codes/2.0/documentation/hummingbird/errorhandling/)
```

                   operationID: String, 
                   next: @Sendable (HTTPTypes.HTTPRequest, OpenAPIRuntime.HTTPBody?, OpenAPIRuntime.ServerRequestMetadata) async throws -> (HTTPTypes.HTTPResponse, OpenAPIRuntime.HTTPBody?)) async throws -> (HTTPTypes.HTTPResponse, OpenAPIRuntime.HTTPBody?) {
        do {
            return try await next(request, body, metadata)
        } catch let error as ServerError {
            if let appError = error.underlyingError as? HTTPStatusConvertible else {
                return (HTTPResponse(status: appError.httpStatus), nil)
            } else if ... {
              
            } 
        }
    }
}
```

Please note that the proposal places the responsibility to conform to the documented API in the hands of the user.
There's no mechanism to prevent the users from inadvertently transforming a thrown error into an undocumented response. 

#### Example usage

1. Create an error type that conforms to one of the error protocol(s)
```swift
extension MyAppError: HTTPStatusConvertible {
    var httpStatus: HTTPResponse.Status {
        switch self {
        case .invalidInputFormat:

>             HTTPResponse.Status.badRequest
>         case .authorizationError:
>             HTTPResponse.Status.forbidden

Reviewer: Nit: for clarity, we can shorten this

```suggestion
            .badRequest
        case .authorizationError:
            .forbidden
```

        }
    }
}
```

2. Opt in to the error middleware while registering the handler

```swift
let handler = try await RequestHandler()

> try handler.registerHandlers(on: transport, middlewares: [ErrorHandlingMiddleware()])

Reviewer: Nit: you're using the name `ErrorMiddleware` earlier, and `ErrorHandlingMiddleware` here. While the name itself will be discussed as part of the review, it might make sense to make them consistent to avoid confusion of whether the names will or will not be the same.

# SOAR-0011: Improved Error Handling

Improve error handling by adding the ability for mapping application errors to HTTP responses.

## Overview

- Proposal: SOAR-0011
- Author(s): [Gayathri Sairamkrishnan](https://github.com/gayathrisairam)

> - Status: **In Review**

Reviewer: ```suggestion
- Status: **Ready for Implementation**
```

> - Status: **Ready for Implementation**

Reviewer: ```suggestion
- Status: **Implemented**
```
